File: OC_TencentVideoChannel/OC_TencentVideoChannel/ViewController.m
//
//  ViewController.m
//  OC_TencentVideoChannel
//
//  Created by 欧阳荣 on 2018/7/18.
//  Copyright © 2018 欧阳荣. All rights reserved.
//

#import "ViewController.h"

#import "ORChannelCell.h"
#import "ORNormalChannelHeader.h"
#import "ORMyChannelHeader.h"
#import "ORScrollMenuView.h"

#import "ORChannelViewModel.h"

static CGFloat const menuHeight = 55.0;
#define Tint_Back_Color [UIColor colorWithRed:245/255.f green:244/255.f blue:247/255.f alpha:1]


@interface ViewController ()<UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout> {
    BOOL _isDraging;
    CGFloat _lastBottowInset;
}
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (nonatomic, strong) ORScrollMenuView *menuView;

@property (nonatomic, strong) ORChannelViewModel *viewModel;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    _viewModel = [ORChannelViewModel new];
    
    self.view.backgroundColor = Tint_Back_Color;
    
    CGFloat menuY = [UIScreen mainScreen].bounds.size.height > 800 ? 75 + 24 : 75;

    [self.collectionView registerNib:[UINib nibWithNibName:@"ORNormalChannelHeader" bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:NSStringFromClass([ORNormalChannelHeader class])];
    [self.collectionView registerClass:[UICollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:NSStringFromClass([UICollectionReusableView class])];

    UILongPressGestureRecognizer *longPressGesture = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(_or_actionLongPressGesture:)];
    [_collectionView addGestureRecognizer:longPressGesture];
    
    
    self.menuView = [ORScrollMenuView new];
    self.menuView.frame = CGRectMake(0, menuY, [UIScreen mainScreen].bounds.size.width, menuHeight);
    self.menuView.titles = _viewModel.titles;
    self.menuView.backgroundColor = Tint_Back_Color;
    self.menuView.hidden = YES;
    [self.view addSubview:self.menuView];

    __weak typeof(self) weakSelf = self;
    [_menuView setMenuDidSelectConfig:^(NSInteger index) {
        [weakSelf _or_collectionViewOffsetAjustMenuWithIndex:index];
    }];
    
}

#pragma mark -- private
- (void)_or_actionLongPressGesture:(UILongPressGestureRecognizer *)gesture {
    
    CGPoint point = [gesture locationInView:_collectionView];
    NSIndexPath *indexPath = [self.collectionView indexPathForItemAtPoint:point];
    
    switch (gesture.state) {
            
        case UIGestureRecognizerStateBegan:
            [self.collectionView beginInteractiveMovementForItemAtIndexPath:indexPath];
            break;
            
        case UIGestureRecognizerStateChanged:
            [self.collectionView updateInteractiveMovementTargetPosition:point];
            break;
            
        case UIGestureRecognizerStateEnded:
            [self.collectionView endInteractiveMovement];
            break;
            
        default:
            [self.collectionView cancelInteractiveMovement];
            break;
    }
}

- (void)_or_collectionViewOffsetAjustMenuWithIndex:(NSInteger)index {
    
    self->_isDraging = NO;
    
    UICollectionViewLayoutAttributes *attr = [self.collectionView layoutAttributesForSupplementaryElementOfKind:UICollectionElementKindSectionHeader atIndexPath:[NSIndexPath indexPathForItem:0 inSection:index + 1]];
    
    CGFloat offsetY = index == 0 ? attr.frame.origin.y : attr.frame.origin.y - menuHeight;
    [self.collectionView setContentOffset:CGPointMake(0,offsetY) animated:YES];
}

#pragma mark -- UIScrollViewDelegate
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    
    if ((_collectionView.visibleCells.count == 0)) {
        return;
    }

    UICollectionViewLayoutAttributes *menuAttr = [self.collectionView layoutAttributesForSupplementaryElementOfKind:UICollectionElementKindSectionHeader atIndexPath:[NSIndexPath indexPathForItem:0 inSection:1]];

    if (scrollView.contentOffset.y < menuAttr.frame.origin.y) {
        _menuView.hidden = YES;
        _collectionView.backgroundColor = Tint_Back_Color;
        return;
    }
    
    _collectionView.backgroundColor = [UIColor colorWithRed:241/255.f green:240/255.f blue:246/255.f alpha:1];
    _menuView.hidden = NO;

    
    if (!_isDraging) {
        return;
    }
    
    NSIndexPath *currentIndexPath = [NSIndexPath indexPathForItem:0 inSection:self.menuView.currentIndex + 1];
    UICollectionViewLayoutAttributes *currentAttr = [self.collectionView layoutAttributesForSupplementaryElementOfKind:UICollectionElementKindSectionHeader atIndexPath:currentIndexPath];
    
    
    if (scrollView.contentOffset.y + menuHeight < currentAttr.frame.origin.y) {
        [_menuView or_setSelectIndex:_menuView.currentIndex-1 animated:YES];
        return;
    }
    
    NSIndexPath *nextIndexPath = [NSIndexPath indexPathForItem:0 inSection:self.menuView.currentIndex + 2];
    if (nextIndexPath.section > _viewModel.dataSource.count-1) {
        return;
    }
    UICollectionViewLayoutAttributes *nextAttr = [self.collectionView layoutAttributesForSupplementaryElementOfKind:UICollectionElementKindSectionHeader atIndexPath:nextIndexPath];
    if (nextAttr && scrollView.contentOffset.y + menuHeight > nextAttr.frame.origin.y) {
        [_menuView or_setSelectIndex:_menuView.currentIndex+1 animated:YES];
    }
}

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
    _isDraging = YES;
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    _isDraging = NO;
}


#pragma mark -- UICollectionViewDelegate && UICollectionViewDataSource

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return _viewModel.dataSource.count;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    
    return _viewModel.dataSource[section].chanels.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    ORChannelCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([ORChannelCell class]) forIndexPath:indexPath];
    cell.titleL.text = _viewModel.dataSource[indexPath.section].chanels[indexPath.item];
    return cell;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
    
    if ([kind isEqualToString:UICollectionElementKindSectionFooter]) {
        
        return [UICollectionReusableView new];
    }
    
    if (indexPath.section == 0) {
        ORMyChannelHeader *header = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:NSStringFromClass([ORMyChannelHeader class]) forIndexPath:indexPath];
        
        return header;
    }
    
    if (indexPath.section == 1) {
        UICollectionReusableView *header = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:NSStringFromClass([UICollectionReusableView class]) forIndexPath:indexPath];
        ORScrollMenuView *menu = [header viewWithTag:2019];
        if (!menu) {
            menu = [[ORScrollMenuView alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, menuHeight)];
            menu.titles = _viewModel.titles;
            menu.tag = 2019;
            [header addSubview:menu];
            __weak typeof(self) weakSelf = self;
            [menu setMenuDidSelectConfig:^(NSInteger index) {
                __strong typeof(weakSelf) strongSelf = weakSelf;
                [strongSelf _or_collectionViewOffsetAjustMenuWithIndex:index];
                [strongSelf.menuView or_setSelectIndex:index animated:YES];
            }];
        }
        [menu or_setSelectIndex:_menuView.currentIndex animated:NO];
        return header;
    }
    
    ORNormalChannelHeader *header = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:NSStringFromClass([ORNormalChannelHeader class]) forIndexPath:indexPath];
    header.titleL.text = _viewModel.dataSource[indexPath.section].title;
    return header;
}

- (BOOL)collectionView:(UICollectionView *)collectionView canMoveItemAtIndexPath:(NSIndexPath *)indexPath {
    return indexPath.section == 0;
}

- (void)collectionView:(UICollectionView *)collectionView moveItemAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath *)destinationIndexPath {
    
    [self.viewModel or_moveItemAtIndexPath:sourceIndexPath.item toIndexPath:destinationIndexPath.item];
}

#pragma mark -- UICollectionViewDelegateFlowLayout

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    
    if (section == 0) {
        return UIEdgeInsetsMake(0, H_P(15), H_P(15), H_P(15));
    }
    if (section == 1) {
        return UIEdgeInsetsMake(H_P(15), H_P(15), 0, H_P(15));
    }
    
    if (section == _viewModel.dataSource.count - 1) {
        return UIEdgeInsetsMake(0, H_P(15), 10, H_P(15));
    }
    
    return UIEdgeInsetsMake(0, H_P(15), 0, H_P(15));
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    if (section == 0) {
        return CGSizeMake(0, 80);
    }
    if (section == 1) {
        return CGSizeMake(0, menuHeight);
    }
    return CGSizeMake(0, 50);
}



@end

File: Swift_TencentVideoChannel/TencentVideoChannel/ViewController.swift
//
//  ViewController.swift
//  TencentVideoChannel
//
//  Created by 欧阳荣 on 2018/7/17.
//  Copyright © 2018 欧阳荣. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {

    @IBOutlet weak var collectionView: UICollectionView!
    
    private let tintColor = UIColor.init(red: 245.0/255.0, green: 244.0/255.0, blue: 247.0/255.0, alpha: 1)
    private let menuHeight:CGFloat = 50.0;
    
    private let menuIdentifer = "menuChannelIdf"
    private let normalIdentifer = "normalChannelIdf"
    
    private let viewModel:ORChannelViewModel = ORChannelViewModel()
    
    private var menuView :ORScrollMenuView!
    
    var isDraging = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.view.backgroundColor = tintColor
        
        collectionView.register(UINib.init(nibName: "ORNormalChannelHeader", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, withReuseIdentifier: normalIdentifer)
        collectionView.register(UICollectionReusableView.self, forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, withReuseIdentifier: menuIdentifer)

        let gesture = UILongPressGestureRecognizer.init(target: self, action: #selector(_or_actionLongPressGesture(gesture:)))
        collectionView.addGestureRecognizer(gesture);
//        Selector.init(stringLiteral: _or)
        
        let menuY:CGFloat = UIScreen.main.bounds.height > 800 ? 75 + 24 : 75
        
        menuView  = ORScrollMenuView(frame: CGRect(x: 0, y: menuY, width: self.view.bounds.width, height: menuHeight))
        menuView.titles = viewModel.titles;
        menuView.backgroundColor = tintColor
        
        self.view.addSubview(menuView)
        menuView.isHidden = true;
        
        menuView.menuDidSelectedClosure = {[unowned self](index:Int) in
            self.collectionViewOffsetAjustMenuWithIndex(index: index)
        }
        
    }

    @objc private func _or_actionLongPressGesture(gesture:UILongPressGestureRecognizer) {
        
        let point = gesture.location(in: collectionView);
        let indexPath = collectionView.indexPathForItem(at: point)
        
        if indexPath == nil {
            return;
        }
        
        switch gesture.state {
        case .began:
            collectionView.beginInteractiveMovementForItem(at: indexPath!)
        case .changed:
            collectionView.updateInteractiveMovementTargetPosition(point)
        case .ended:
            collectionView.endInteractiveMovement()
        default:
            collectionView.cancelInteractiveMovement()
        }
    }
    
    private func collectionViewOffsetAjustMenuWithIndex(index:Int) {
        
        isDraging = false;
        
        let indexPath = IndexPath(item: 0, section: index + 1)
        let attr = self.collectionView.layoutAttributesForSupplementaryElement(ofKind: UICollectionElementKindSectionHeader, at: indexPath)
        let offsetY = indexPath.section == 1 ? attr?.frame.minY : (attr?.frame.minY)! - self.menuHeight
        
        self.collectionView.setContentOffset(CGPoint(x: 0, y: offsetY!), animated: true)
    }
    
    //MARK: - UICollectionViewDataSource && UICollectionViewDelegate
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        
        return viewModel.dataSource[section].chanels.count
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return viewModel.dataSource.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell:ORChannelCell = collectionView.dequeueReusableCell(withReuseIdentifier: "channelCellIdf", for: indexPath) as! ORChannelCell
        cell.titleL.text = viewModel.dataSource[indexPath.section].chanels[indexPath.item]
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionElementKindSectionHeader {
            
            if indexPath.section == 0 {
                let header:ORMyChannelHeader = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "myChannelIdf", for: indexPath) as! ORMyChannelHeader;
                return header;
            }
            
            if indexPath.section == 1 {
                let header = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: menuIdentifer, for: indexPath);
                
                var view:ORScrollMenuView? = header.viewWithTag(2019) as? ORScrollMenuView;
                if view == nil {
                    view = ORScrollMenuView(frame: CGRect(x: 0, y: 0, width: self.view.bounds.width, height: menuHeight))
                    view?.titles = viewModel.titles;
                    view?.tag = 2019;
                    header.addSubview(view!);
                    view?.menuDidSelectedClosure = {[unowned self](index:Int) in
                        self.collectionViewOffsetAjustMenuWithIndex(index: index)
                        self.menuView.or_setSelectedIndex(index: index, animated: true)
                    }
                }
                
                view?.or_setSelectedIndex(index: menuView.currentIndex, animated: false)
                
                return header;
            }
            
            let header:ORNormalChannelHeader = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: normalIdentifer, for: indexPath) as! ORNormalChannelHeader;
            header.titleL.text = viewModel.dataSource[indexPath.section].title
            return header;
            
        }
        return UICollectionReusableView()
    }
    
    func collectionView(_ collectionView: UICollectionView, canMoveItemAt indexPath: IndexPath) -> Bool {
        return indexPath.section == 0 && viewModel.canMove == true
    }
    
    func collectionView(_ collectionView: UICollectionView, moveItemAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {
        viewModel.or_moveItemAtIndex(sourceIndex: sourceIndexPath.item, destinationIndex: destinationIndexPath.item)
    }
    
    //MARK: - UIScrollViewDelegate

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        
        
        if collectionView.visibleCells.count == 0 {
            return;
        }
        
        let indexPath = IndexPath(item: 0, section: 1);
        let attr = collectionView.layoutAttributesForSupplementaryElement(ofKind: UICollectionElementKindSectionHeader, at: indexPath)
        
        if scrollView.contentOffset.y < ((attr?.frame.origin.y)! ) {
            menuView.isHidden = true;
            collectionView.backgroundColor = tintColor;
            return;
        }
        
        collectionView.backgroundColor = UIColor.init(red: 241.0/255.0, green: 240.0/255.0, blue: 246.0/255.0, alpha: 1)
        menuView.isHidden = false;
        
        if isDraging == false {
            return;
        }
        
        let currentIndexPath = IndexPath(item: 0, section: menuView.currentIndex + 1);
        let currentAttr = collectionView.layoutAttributesForSupplementaryElement(ofKind: UICollectionElementKindSectionHeader, at: currentIndexPath)
        
        if scrollView.contentOffset.y + menuHeight < (currentAttr?.frame.minY)! {
            menuView.or_setSelectedIndex(index: menuView.currentIndex - 1, animated: true);
        }else {
            let nextIndexPath = IndexPath(item: 0, section: menuView.currentIndex + 2);
            
            if nextIndexPath.section > viewModel.dataSource.count-1 {
                return;
            }
            
            let nextAttr = collectionView.layoutAttributesForSupplementaryElement(ofKind: UICollectionElementKindSectionHeader, at: nextIndexPath)
            if nextAttr != nil && scrollView.contentOffset.y + menuHeight > (nextAttr?.frame.minY)!{
                menuView.or_setSelectedIndex(index: menuView.currentIndex + 1, animated: true)
            }
        }
    }

    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        isDraging = true
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        isDraging = false;
    }
    
    //MARK: - UICollectionViewDelegateFlowLayout
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        if section == 0 {
            return CGSize(width: 0, height: 80);
        }
        if section == 1 {
            return CGSize(width: 0, height: menuHeight);
        }
        return CGSize(width: 0, height: 50);
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        
        let insetLeft = 15.0 * H_PRATIO
        
        if section == 0 {
            return UIEdgeInsetsMake(0, insetLeft, insetLeft, insetLeft)
        }
        if section == 1 {
            
            return UIEdgeInsetsMake(insetLeft, insetLeft, 0, insetLeft)
        }
        
        if section == viewModel.dataSource.count - 1 {
            return UIEdgeInsetsMake(0, insetLeft, 10, insetLeft)
        }
        
        return UIEdgeInsetsMake(0, insetLeft, 0, insetLeft)
    }

}

